plugins/permission-backend/src/schema/openapi.yaml

```diff
@@ -0,0 +1,296 @@
+                        type: string
+                      attributes:
+                        type: object
+                        properties:
+                          action:
+                            type: string
+                            enum:
+                              - create
+                              - read
+                              - update
+                              - delete
+                        additionalProperties: false
+                      type:
+                        type: string
+                        enum: 
+                          - basic
+                    required:
+                      - name
+                      - type
+                      - attributes
+                    additionalProperties: false
+                  - type: object
+                    properties:
+                      type:
+                        type: string
+                        enum: 
+                          - resource
+                      resourceType:
+                        type: string
+                      name:
+                        type: string
+                      attributes:
+                        type: object
+                        properties:
+                          action:
+                            type: string
+                            enum:
+                              - create
+                              - read
+                              - update
+                              - delete
+                        additionalProperties: false
+                    required:
+                      - type
+                      - name
+                      - attributes
+                      - resourceType
+                    additionalProperties: false
+              resourceRef:
+                type: string
+            required:
+              - id
+              - permission
+            additionalProperties: false
+      required:
+        - items
+    
+    PermissionRuleParam:
+      oneOf:
+        - type: string
+        - type: number
+        - type: boolean
+          nullable: true
+
+    PermissionCondition:
+      type: object
+      properties:
+        resourceType:
+          type: string
+        rule:
+          type: string
+        params:
+          type: object
+          additionalProperties:
+            oneOf:
+              - $ref: "#/components/schemas/PermissionRuleParam"
+              - type: array
+                items:
+                  $ref: "#/components/schemas/PermissionRuleParam"
+      required:
+        - resourceType
+        - rule
+      additionalProperties: true
+
+    DefinitivePolicyDecision:
+      type: object
+      properties:
+        result:
+          type: string
+          enum:
+            - ALLOW
+            - DENY
+      required:
+        - result
+      additionalProperties: false
+
+    PermissionCriteria:
+      # Restricting this to depth of 1, TS doesn't love recursive
```

Comment:
Indeterminate recursive types don't resolve too well with the json schema library we're using. Restricting this to depth 1 loses some typings, but makes the types able to resolve.

Reply:
:'(

Reply:
Does this still validate > 1 depth?

Reply:
This ones just a response type so it wouldn't need to validate. If we added a validation endpoint, it would be a bit different. I stumbled upon https://github.com/geut/openapi-box + TypeBox which might be a better direction to take some of the code generation if we need recursion.